File: src/MyServiceBus/Topology/ConsumerDefinitionModel.cs
namespace MyServiceBus.Topology;

/// <summary>
/// Immutable consumer policy captured before topology materialization.
/// </summary>
public sealed record ConsumerDefinitionModel
{
    public ConsumerDefinitionModel(
        Type consumerType,
        string endpointName,
        bool endpointNameIsExplicit,
        Type? endpointNameFormatterType,
        IEnumerable<Type> messageTypes,
        int? concurrentMessageLimit)
    {
        ArgumentNullException.ThrowIfNull(consumerType);
        ArgumentException.ThrowIfNullOrWhiteSpace(endpointName);
        ArgumentNullException.ThrowIfNull(messageTypes);
        if (concurrentMessageLimit is <= 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(concurrentMessageLimit),
                concurrentMessageLimit,
                "The concurrent message limit must be greater than zero.");
        }

        var capturedMessageTypes = messageTypes.Distinct().ToArray();
        if (capturedMessageTypes.Length == 0)
            throw new ArgumentException("At least one consumed message type is required.", nameof(messageTypes));
        if (capturedMessageTypes.Any(messageType => messageType is null))
            throw new ArgumentException("Consumed message types must not contain null.", nameof(messageTypes));

        ConsumerType = consumerType;
        EndpointName = endpointName;
        EndpointNameIsExplicit = endpointNameIsExplicit;
        EndpointNameFormatterType = endpointNameFormatterType;
        MessageTypes = capturedMessageTypes;
        ConcurrentMessageLimit = concurrentMessageLimit;
    }

    public Type ConsumerType { get; }

    public string EndpointName { get; }

    public bool EndpointNameIsExplicit { get; }

    public Type? EndpointNameFormatterType { get; }

    public IReadOnlyList<Type> MessageTypes { get; }

    public int? ConcurrentMessageLimit { get; }
}

File: src/Java/myservicebus/src/main/java/com/myservicebus/topology/ConsumerDefinitionModel.java
package com.myservicebus.topology;

import java.util.List;

/** Immutable consumer policy captured before topology materialization. */
public record ConsumerDefinitionModel(
        Class<?> consumerType,
        String endpointName,
        boolean endpointNameExplicit,
        Class<?> endpointNameFormatterType,
        List<Class<?>> messageTypes,
        Integer concurrentMessageLimit) {
    public ConsumerDefinitionModel {
        if (consumerType == null) {
            throw new IllegalArgumentException("consumerType must not be null");
        }
        if (endpointName == null || endpointName.isBlank()) {
            throw new IllegalArgumentException("Endpoint name must not be blank.");
        }
        if (messageTypes == null || messageTypes.isEmpty()) {
            throw new IllegalArgumentException("At least one consumed message type is required.");
        }
        if (messageTypes.stream().anyMatch(java.util.Objects::isNull)) {
            throw new IllegalArgumentException("Consumed message types must not contain null.");
        }
        if (concurrentMessageLimit != null && concurrentMessageLimit < 1) {
            throw new IllegalArgumentException("Concurrent message limit must be greater than zero.");
        }
        messageTypes = List.copyOf(messageTypes.stream().distinct().toList());
    }
}
